docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,401 @@
+# ADR 082: Data Companion API
+
+## Changelog
+
+- 2022-09-10: First draft (@thanethomson)
+
+## Status
+
+Accepted | Rejected | Deprecated | Superseded by
+
+## Context
+
+This ADR proposes to effectively offload some data storage responsibilities and
+functionality to **single external companion service** in order to:
+
+1. Improve core consensus stability.
+2. Eventually reduce the amount of functionality for which the core team is
+   responsible, thereby improving maintainability.
+3. Still cater for certain use cases that require this data and functionality.
+
+The way the system is currently built is such that a Tendermint node is mostly
+self-contained. While this philosophy initially allowed a certain degree of ease
+of node operation (i.e. simple UX), it has also lent itself to feature sprawl,
+with Tendermint being asked to take care of increasingly more than just
+consensus. Under the spotlight in this ADR are:
+
+1. **Event indexing**, which is required in order to facilitate arbitrary block
+   and transaction querying, as well as subscription for arbitrary events. We
+   have already seen, for instance, how the current (somewhat unreliable) event
+   subscription implementation can cause back-pressure into consensus and affect
+   IBC relayer stability (see [\#6729] and [\#7156]).
+2. **Block execution result storage**, which is required to facilitate a number
+   of RPC APIs, but also results in storage of large quantities of data that is
+   not critical to consensus. This data is, however, critical for certain use
+   cases outside of consensus.
+
+Another intersecting issue is that there may be a variety of use cases that
+require different data models. A good example of this is the mere existence of
+the [PostgreSQL indexer] as compared to the default key/value indexer.
```

Comment:
Is it reasonable to expect Tendermint to directly accommodate every use case with sufficient demand?

Reply:
Not necessarily. If there's sufficient demand for functionality that we consider to be outside of "core" Tendermint functionality, then we should find some way of accommodating that, but that doesn't mean we need to implement and maintain it ourselves.

This is what this ADR attempts to do.

We're a small team with limited capacity. In order for us to meet our mid- to long-term correctness goals for the system, we believe we need to find ways in the short term to trim away and outsource certain parts of the system that we believe are not core to the system's operations - especially parts of the system where Tendermint's effectively catering for application-level concerns.